### PR TITLE
Update the definition of trainable params with Jax: right trainable parameters for JIT

### PR DESCRIPTION
**Pull request by rmoyard**
**Context:**
This PR adresses two issues:

1. All jitted parameters are considered trainable and therefore we always compute too many jacobians.
2. Argnums in transforms to set trainable params (Hessian tranform, batch_params and batch_inputs) See https://github.com/PennyLaneAI/pennylane/pull/3750

**Description of the Change:**

We change the definition of trainable parameters for Jax; a trainable parameters in Jax is `jax.interpreters.ad.JVPTracer`.

We update the new return type pipeline to take advantage of the  correct definition of trainable parameter and compute less jacobians. The old return pipeline is not updated but adapted to the change and still compute the jacobians for too many parameters.

The hessian transform is adapted to be able to define trainable parameters with `argnum`.

**Benefits:**

Computational cost of using jitting with derivative is reduced. We move towards becoming closer to a parity of feature for the Jax interface.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/gradients/parameter_shift_hessian.py [hunk 1]
```diff
@@ -654,6 +654,15 @@ def param_shift_hessian(tape, argnum=None, diagonal_shifts=None, off_diagonal_sh
...
 
+    # For Jax the argnum defines the tape trainable parameters (it does not work for an array argnum).
+    if (
+        argnum is not None
+        and not isinstance(argnum, np.ndarray)
+        and not tape.trainable_params
+        and qml.math.get_interface(*tape.data) in ["jax", "jax-jit"]
+    ):
+        tape.trainable_params = argnum
+
```
Comment (dwierichs) [inline pennylane/gradients/parameter_shift_hessian.py:665]: I'm a bit worried about this change:
- Newly added similar transforms will have to include this, but it is not clear that this step is necessary, or when this needs to be done.
- It manually tracks trainability, which feels like we should not have to do.

Why is this kind of change not required with gradient transforms? :thinking:

File: pennylane/interfaces/jax_jit.py [hunk 2]
```diff
@@ -73,10 +73,15 @@ def execute(tapes, device, execute_fn, gradient_fn, gradient_kwargs, _n=1, max_d
...
-        params = tape.get_parameters(trainable_only=False)
-        tape.trainable_params = qml.math.get_trainable_indices(params)
+    if _n == 1:
+        for tape in tapes:
+            # set the jitted parameters
+            params = tape.get_parameters(trainable_only=False)
+            trainable_params = set()
+            for idx, p in enumerate(params):
+                if isinstance(p, jax.core.Tracer) or qml.math.requires_grad(p):
+                    trainable_params.add(idx)
```
Comment (dwierichs) [inline pennylane/interfaces/jax_jit.py:83]: I believe `trainable_params` is supposed to be a `list`, because we require it to maintain order (and be sorted).

```suggestion
            trainable_params = []
            for idx, p in enumerate(params):
                if isinstance(p, jax.core.Tracer) or qml.math.requires_grad(p):
                    trainable_params.append(idx)

```

File: pennylane/interfaces/jax_jit_tuple.py [hunk 3]
```diff
@@ -95,6 +95,18 @@ def _jac_shape_dtype_tuple(tapes, device):
     return tuple(shape_dtypes)
 
 
+def _jit_compute_jvps(jacobians, tangents, multi_measurements, trainable_params):
+    # Remove the the jitted parameters that are no trainable as it is not contributing to the jacobian.
```
Comment (dwierichs) [inline pennylane/interfaces/jax_jit_tuple.py:99]:
```suggestion
    # Remove the the jitted parameters that are not trainable as they are not contributing to the jacobian.
```

File: pennylane/interfaces/jax_jit_tuple.py [hunk 4]
```diff
@@ -95,6 +95,18 @@ def _jac_shape_dtype_tuple(tapes, device):
...
+    tangents_trainable = []
+    for i, tangent in enumerate(tangents):
+        tangent_ = [tangent[j] for j in trainable_params[i]]
+        tangents_trainable.append(tangent_)
```
Comment (dwierichs) [inline pennylane/interfaces/jax_jit_tuple.py:103]: Could do this:
```suggestion
tangents_trainable = [[tangent[idx] for idx in trainable] for tangent, trainable in zip(tangents, trainable_params)]
```

Comment (dwierichs) [reply pennylane/interfaces/jax_jit_tuple.py:103]: Or just maintain the loop structure but using `zip` for it.

File: pennylane/interfaces/jax_jit.py [hunk 5]
```diff
@@ -73,10 +73,15 @@ def execute(tapes, device, execute_fn, gradient_fn, gradient_kwargs, _n=1, max_d
     if max_diff > 1:
         raise InterfaceUnsupportedError("The JAX interface only supports first order derivatives.")
 
-    for tape in tapes:
-        # set the trainable parameters
-        params = tape.get_parameters(trainable_only=False)
-        tape.trainable_params = qml.math.get_trainable_indices(params)
+    if _n == 1:
```
Comment (dwierichs) [inline pennylane/interfaces/jax_jit.py:76]: So if the order of the derivative is higher, we do not need to update the trainable parameters? Why is this? :)

File: pennylane/interfaces/jax_jit_tuple.py [hunk 6]
```diff
@@ -203,6 +218,10 @@ def wrapper(p):
     def execute_wrapper_jvp(primals, tangents):
         # pylint: disable=unused-variable
         params = primals[0]
+        # Select the trainable params
+        # Non trainable params are 0
```
Comment (dwierichs) [inline pennylane/interfaces/jax_jit_tuple.py:222]: I do not quite understand this comment, sorry :grimacing: What exactly is 0, and how does it relate to the following code line?

File: pennylane/math/utils.py [hunk 7]
```diff
@@ -474,7 +474,7 @@ def requires_grad(tensor, interface=None):
     if interface == "jax":
         import jax
 
-        return isinstance(tensor, jax.core.Tracer)
+        return isinstance(tensor, jax.interpreters.ad.JVPTracer)
```
Comment (dwierichs) [inline pennylane/math/utils.py:477]: :100:

File: pennylane/math/multi_dispatch.py [hunk 8]
```diff
@@ -410,28 +410,10 @@ def get_trainable_indices(values, like=None):
...
-
-        else:
-            # JAX tracing is occuring; use the default behaviour (only traced arrays
-            # are treated as trainable). This is required to ensure that `jax.grad(func, argnums=...)
-            # works correctly, as the argnums argnument determines which parameters are
-            # traced arrays.
-            trainable = requires_grad
-
     for idx, p in enumerate(values):
-        if trainable(p, interface=like):
```
Comment (dwierichs) [inline pennylane/math/multi_dispatch.py:434]: This looks so much cleaner now, amazing!

File: pennylane/ops/qubit/parametric_ops.py [hunk 9]
```diff
@@ -40,7 +40,7 @@ def _can_replace(x, y):
     Convenience function that returns true if x is close to y and if
     x does not require grad
     """
-    return (not qml.math.requires_grad(x)) and qml.math.allclose(x, y)
+    return not qml.math.is_abstract(x) and not qml.math.requires_grad(x) and qml.math.allclose(x, y)
```
Comment (dwierichs) [inline pennylane/ops/qubit/parametric_ops.py:43]: Why is this needed? Is there a rule of thumb in which cases user code might require changes like this?

File: tests/interfaces/test_jax_qnode.py [hunk 10]
```diff
@@ -73,11 +73,16 @@ def circuit(a):
             return qml.expval(qml.PauliZ(0))
 
         a = np.array(0.1, requires_grad=True)
-        circuit(a)
+
+        if interface == "jax-jit":
+            jax.jit(circuit)(a)
+        else:
+            circuit(a)
```
Comment (dwierichs) [inline tests/interfaces/test_jax_qnode.py:80]: Wait, were we not testing JIT for `interface="jax-jit"`?

File: tests/returntypes/transforms/test_batch_params_new.py [hunk 11]
```diff
@@ -487,7 +487,7 @@ def test_jax(self, diff_method, tol):
 
         dev = qml.device("default.qubit", wires=2)
 
-        @qml.batch_params
+        @qml.batch_params(all_operations=True)
```
Comment (dwierichs) [inline tests/returntypes/transforms/test_batch_params_new.py:490]: Just out of curiosity: How does this change come about? :D 
Is it something users have to change as well?

**Review by dwierichs — COMMENTED** [review]
Very nice to see the updated definition of `requires_grad` and the simpler `get_trainable_indices`! :100: Thanks @rmoyard :)

I'm wondering about breaking changes, though, as there are a few detached places in the codebase and tests that required changes:
- The need to provide `argnum` to the hessian transform
- Having to set `all_operations=True` in `batch_params` (or why is it changed in the tests?)
- Sometimes updating the trainable params of a tape seems to be required in addition?

 Are there breaking changes?
In any case, could you also add a changelog entry? :)

Otherwise I left a few minor suggestions/comments.

**Comment by rmoyard** [comment]
[sc-31370]

File: pennylane/interfaces/jax_jit.py (hunk 2 above)
Comment (rmoyard) [reply pennylane/interfaces/jax_jit.py:83]: I took this from the math module where we define the following function:

```
@multi_dispatch(argnum=[0], tensor_list=[0])
def get_trainable_indices(values, like=None):
    trainable_params = set()

    for idx, p in enumerate(values):
        if requires_grad(p, interface=like):
            trainable_params.add(idx)

    return trainable_params
```

File: pennylane/interfaces/jax_jit_tuple.py (hunk 6 above)
Comment (rmoyard) [reply pennylane/interfaces/jax_jit_tuple.py:221]: Parameters here are also parameters that are jitted but not trainable, the non-trainable will contribute 0 for the JVP.

File: pennylane/ops/qubit/parametric_ops.py (hunk 9 above)
Comment (rmoyard) [reply pennylane/ops/qubit/parametric_ops.py:43]: Because you are not supposed to be able to compare the value of jitted operations, previsouly a jitted parameter would return `requires_grad=True` if without being trainable, so I had to add the check.

File: tests/interfaces/test_jax_qnode.py (hunk 10 above)
Comment (rmoyard) [reply tests/interfaces/test_jax_qnode.py:80]: That's true just the interface but not jitted.

File: tests/returntypes/transforms/test_batch_params_new.py (hunk 11 above)
Comment (rmoyard) [reply tests/returntypes/transforms/test_batch_params_new.py:490]: Only trainable parameters are considered for this transform by default, previously if all parameters are jax Arrays we would consider trainable, but it is not the case anymore. As batch_input and batch_params are covered by broadcasting, I don't think it is worth adding argnum for Jax.

File: pennylane/interfaces/jax_jit.py (hunk 5 above)
Comment (rmoyard) [reply pennylane/interfaces/jax_jit.py:76]: You always start at _n=1 and only need to fix your trainable parameters there. For hessian for example you will finish at `_n=2` with only jax arrray and nothing is considered trainable and cannot get the jacobian, therefore I needed to add this condition.

File: pennylane/gradients/parameter_shift_hessian.py (hunk 1 above)
Comment (rmoyard) [reply pennylane/gradients/parameter_shift_hessian.py:662]: I think it is part of a bigger change I want to push in separate PRs, users can use argnums to tag trainable parameters in Jax. Because otherwise everything is considered non trainable when you don't apply gradjac/hessian.

File: pennylane/gradients/hessian_transform.py [hunk 12]
```diff
@@ -161,7 +161,10 @@ def default_qnode_wrapper(self, qnode, targs, tkwargs):
         cjac_fn = qml.transforms.classical_jacobian(qnode, expand_fn=self.expand_fn)
 
         def hessian_wrapper(*args, **kwargs):
-            if not qml.math.get_trainable_indices(args):
+            if not qml.math.get_trainable_indices(args) and not qml.math.get_interface(*args) in [
+                "jax",
+                "jax-jit",
+            ]:
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/gradients/hessian_transform.py:167]: Because you want to set your trainable parameters with argnum

**Comment by rmoyard** [comment]
[sc-31369]

File: pennylane/gradients/hessian_transform.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/gradients/hessian_transform.py:173]: If you give only Jax arrays you will get `qml.math.get_trainable_indices(args)=[]`

**Comment by rmoyard** [comment]
Transforms issue arising from this change is fixed there: https://github.com/PennyLaneAI/pennylane/pull/3750

File: pennylane/gradients/hessian_transform.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/gradients/hessian_transform.py:173]: Fix there: https://github.com/PennyLaneAI/pennylane/pull/3750

File: pennylane/interfaces/jax_jit.py (hunk 2 above)
Comment (dwierichs) [reply pennylane/interfaces/jax_jit.py:83]: :thinking:

File: tests/returntypes/jax/test_jax_jit_qnode_new.py [hunk 13]
```diff
@@ -70,7 +70,7 @@ def circuit(a):
         assert circuit.interface == interface
 
         # the tape is able to deduce trainable parameters
-        assert circuit.qtape.trainable_params == [0]
+        assert circuit.qtape.trainable_params == []
```
Comment (dwierichs) [inline tests/returntypes/jax/test_jax_jit_qnode_new.py:73]: What was the previous behaviour doing here? Why was the gradient empty if we explicitly set a trainable parameter? Was it ignored because of JITting?

File: tests/returntypes/jax/test_jax_jit_qnode_new.py [hunk 14]
```diff
@@ -287,6 +287,8 @@ def circuit(a, b):
         assert np.allclose(res[1], expected[1], atol=tol, rtol=0)
 
         res = jax.jit(jax.jacobian(circuit, argnums=[0, 1]))(a, b)
+        assert circuit.qtape.trainable_params == [0, 1]
```
Comment (dwierichs) [inline tests/returntypes/jax/test_jax_jit_qnode_new.py:290]: Nice.

**Review by dwierichs — COMMENTED** [review]
I gave it another round, together with the review of #3750.
To me the following questions would be important:
- What are the breaking changes of these two PRs? We definitely should track them and note them in the changelog. They will be close to impossible for users to track down.
One aspect here also is the changing of trainable parameters of tapes under the hood I think.
- Together with #3750, this introduces the keyword argument `argnums` to PennyLane gradient transforms. PL functions use the nomenclature `argnum` instead. What happens if a user passes 1.  `argnum` to a transform while using JAX, and 2. `argnums` to some transform while using a different interface?
- Are there any performance implications with `jax_argnums_to_tape_trainable` (from #3750)?
- Optimally we could add some comments on how to use `argnum` (or `argnums`) with a `gradient_transform` to the `gradients` module documentation, but that may be done separately, I'd say.

I hope these questions and concerns make sense. I just want to make sure we're not moving to swiftly here, in a terrain in which it might be hard for users to follow.

File: tests/returntypes/jax/test_jax_jit_qnode_new.py (hunk 13 above)
Comment (rmoyard) [reply tests/returntypes/jax/test_jax_jit_qnode_new.py:73]: Jitting would make the first parameter a tracer and therefore we would end up identify it as trainable. But the gradient is not empty.

```
import pennylane as qml
import pennylane.numpy as np

import jax
from jax.config import config

remember = config.read("jax_enable_x64")
config.update("jax_enable_x64", True)

qml.enable_return()
dev = qml.device("default.qubit", wires=2)

@qml.qnode(dev, interface="jax", diff_method="parameter-shift")
def circuit(a):
    qml.RY(a, wires=0)
    qml.RX(0.2, wires=0)
    return qml.expval(qml.PauliZ(0))
a = np.array(0.1, requires_grad=True)

print(jax.jit(circuit)(a))
print(circuit.tape.trainable_params)
```
```
0.9751703272018156
[0]
```

**Comment by rmoyard** [comment]
First in my opinion we should encourage the use of argnums when using Jax. Multiple solutions for that: 
1. We update the documentation and encourage using it. 
2. We raise a warning
3. Internally we can switch argnum to argnums for Jax (temporarly)

My plan is to add all the breaking changes in the changelog, it includes:

1. What is a trainable parameters with JAX,
2.  How setting argnums dynamically changes the trainable parameters
4. How to use argnums with PennyLane transforms and Jax

In the future we should update the Jax interface page to reflect all those changes and guide the user with example

Performance wise I don't think the regression is too big: see #3750 benchmark.

File: pennylane/interfaces/jax_jit.py [hunk 15]
```diff
@@ -73,10 +73,15 @@ def execute(tapes, device, execute_fn, gradient_fn, gradient_kwargs, _n=1, max_d
...
-        # set the trainable parameters
-        params = tape.get_parameters(trainable_only=False)
-        tape.trainable_params = qml.math.get_trainable_indices(params)
+    if _n == 1:
+        for tape in tapes:
+            # set the jitted parameters
+            params = tape.get_parameters(trainable_only=False)
+            trainable_params = set()
+            for idx, p in enumerate(params):
+                if isinstance(p, jax.core.Tracer) or qml.math.requires_grad(p):
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/interfaces/jax_jit.py:82]: It actually breaks the following code, by setting the non-requested, non-differentiable parameter `y` to be a trainable tape parameter:
```python
import jax
import pennylane as qml
jnp = jax.numpy

dev = qml.device("default.qubit", wires=1, shots=None)
    
class MyRY(qml.RY):
    """Just an RY copy that is not diffable"""
    grad_method = None

@qml.qnode(dev, interface="jax", diff_method="parameter-shift")
def circuit(x, y):
    MyRY(y, 0)
    qml.Rot(*x, 0)
    return qml.expval(qml.PauliZ(0))

x = jnp.array([0.4, 0.2, 1.2])
y = 0.7

grad_fn = jax.jit(jax.grad(circuit, argnums=0))
grad_fn(x, y) # breaks, reporting that the tape parameter with index 0 is trainable, which is not supported by MyRY
```

When removing the `if isinstance(p, jax.core.Tracer)` part, this error is no longer raise, but we get a leaked tracer error:
```python
---------------------------------------------------------------------------
XlaRuntimeError                           Traceback (most recent call last)
Cell In[2], line 22
     19 y = 0.7
     21 grad_fn = jax.jit(jax.grad(circuit, argnums=0))
---> 22 grad_fn(x, y)

    [... skipping hidden 3 frame]

File ~/venvs/dev/lib/python3.10/site-packages/jax/interpreters/pxla.py:2113, in ExecuteReplicated.__call__(self, *args)
   2111   out_bufs = self._call_with_tokens(input_bufs)
   2112 else:
-> 2113   out_bufs = self.xla_executable.execute_sharded_on_local_devices(
   2114       input_bufs)
   2115 if dispatch.needs_check_special():
   2116   for bufs in out_bufs:

XlaRuntimeError: INTERNAL: Generated function failed: CpuCallback error: UnexpectedTracerError: Encountered an unexpected tracer. A function transformed by JAX had a side effect, allowing for a reference to an intermediate value with type float64[] wrapped in a DynamicJaxprTracer to escape the scope of the transformation.
JAX transformations require that functions explicitly return their outputs, and disallow saving intermediate values to global state.
The function being traced when the value leaked was circuit at /tmp/ipykernel_498342/2766699701.py:12 traced for jit.
------------------------------
The leaked intermediate value was created on line /tmp/ipykernel_498342/2766699701.py:22 (<module>). 
------------------------------
When the value was created, the final 5 stack frames (most recent last) excluding JAX-internal frames were:
------------------------------
/home/david/venvs/dev/lib/python3.10/site-packages/IPython/core/async_helpers.py:129 (_pseudo_sync_runner)
/home/david/venvs/dev/lib/python3.10/site-packages/IPython/core/interactiveshell.py:3194 (run_cell_async)
/home/david/venvs/dev/lib/python3.10/site-packages/IPython/core/interactiveshell.py:3373 (run_ast_nodes)
/home/david/venvs/dev/lib/python3.10/site-packages/IPython/core/interactiveshell.py:3433 (run_code)
/tmp/ipykernel_498342/2766699701.py:22 (<module>)
------------------------------
```

**Review by dwierichs — COMMENTED** [review]
Thanks for taking care of the issues and suggestions in #3750 and answering my questions here.
I have one more question about a particular change to which tape parameters are set to be trainable, tied to an annoying bug :grimacing:

File: pennylane/interfaces/jax_jit.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/interfaces/jax_jit.py:82]: It is because it is the definition in the old interface, a trainable parameter is a Tracer. I made changes in the interface (old) Jax because updating to the new definition (JVPTracer) was cumbersome. I will take a look at this bug, thank you!

File: pennylane/interfaces/jax_jit_tuple.py [hunk 16]
```diff
@@ -95,6 +95,18 @@ def _jac_shape_dtype_tuple(tapes, device):
     return tuple(shape_dtypes)
 
 
+def _jit_compute_jvps(jacobians, tangents, multi_measurements, trainable_params):
+    # Remove the the jitted parameters that are not trainable as they are not contributing to the jacobian.
```
Comment (timmysilv) [inline pennylane/interfaces/jax_jit_tuple.py:99]: none of the other interfaces do this, right? I suppose the user wouldn't know the difference (other than jax-jit being fast), but I'm wondering if there are consequences to this

File: tests/devices/test_default_mixed_jax.py [hunk 17]
```diff
@@ -667,7 +667,7 @@ def test_batching(self, jacobian_fn, decorator, tol):
             # TODO: https://github.com/PennyLaneAI/pennylane/issues/2762
             pytest.xfail("Parameter broadcasting currently not supported for JAX jit")
```
Comment (timmysilv) [inline tests/devices/test_default_mixed_jax.py:668]: unrelated, but this bug is closed as fixed. I removed this xfail locally on master and it still fails - I will try to check it out later, just commenting to get visibility

File: pennylane/interfaces/jax_jit_tuple.py (hunk 16 above)
Comment (timmysilv) [reply pennylane/interfaces/jax_jit_tuple.py]: nvm, this question doesn't really make sense for other interfaces because this is a jit issue 😄

Comment (rmoyard) [reply pennylane/interfaces/jax_jit_tuple.py]: Here the issue is that params are all the jitted one because you need it for the `callback`, but at the same time you want to get the jacobian but only for the trainable ones. Therefore you need to remove them when you get the JVPs. It is only for Jax-jit and only for performance.

File: tests/devices/test_default_mixed_jax.py (hunk 17 above)
Comment (rmoyard) [reply tests/devices/test_default_mixed_jax.py:668]: Good catch let's take a look next week during feature freeze!

File: pennylane/interfaces/jax_jit.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/interfaces/jax_jit.py:82]: I digged this issue;

master and this branch have the same behaviour for the old return type: (wrong behaviour)
```
jaxlib.xla_extension.XlaRuntimeError: INTERNAL: Generated function failed: CpuCallback error: ValueError: Cannot differentiate with respect to parameter(s) {0}
```

this branch and the new return type: (correct behaviour)
```
[ 4.98402115e-02 -7.33486698e-01  1.46230893e-17]
```

This PR aims to do the changes only for the new return type, so I think it is good as it is. Let me know what you think

**Comment by josh146** [comment]
> First in my opinion we should encourage the use of argnums when using Jax. Multiple solutions for that:

Just to double check @rmoyard, will we default to `argnums=[0]` if not provided by the user? This follows JAX I believe

File: pennylane/interfaces/jax_jit.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/interfaces/jax_jit.py:82]: Ah, I understand! Nice, thanks for digging into this. 
Yes, as we do not break previously working code, this definitely should be fine :)

File: tests/devices/test_default_mixed_jax.py (hunk 17 above)
Comment (dwierichs) [reply tests/devices/test_default_mixed_jax.py:668]: :grimacing:

**Comment by rmoyard** [comment]
@josh146 yes exactly we will follow the convention of Jax and Functorch, default being 0.

File: pennylane/transforms/classical_jacobian.py [hunk 18]
```diff
@@ -184,16 +184,6 @@ def _jacobian(*args, **kwargs):  # pylint: disable=unused-argument
             argnum = 0 if wrapper_argnum is None else wrapper_argnum
 
             def _jacobian(*args, **kwargs):
-                if trainable_only:
```
Comment (timmysilv) [inline pennylane/transforms/classical_jacobian.py:187]: confused how this worked before.. `trainable_only` doesn't seem to be defined in this scope

**Review by timmysilv — COMMENTED** [review]
looking good! I think it's pretty much ready to go

File: pennylane/gradients/gradient_transform.py [hunk 19]
```diff
@@ -277,19 +277,22 @@ def default_qnode_wrapper(self, qnode, targs, tkwargs):
...
         # inside the QNode.
-        hybrid = tkwargs.pop("hybrid", self.hybrid)
-        argnums = tkwargs.pop("argnums", None)
+        argnums = tkwargs.get("argnums", None)
 
         _wrapper = super().default_qnode_wrapper(qnode, targs, tkwargs)
+        cjac_fn = qml.transforms.classical_jacobian(qnode, argnum=argnums, expand_fn=self.expand_fn)
 
-        cjac_fn = qml.transforms.classical_jacobian(
-            qnode, argnum=argnums, expand_fn=expand_invalid_trainable
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/gradients/gradient_transform.py:286]: I think they were assuming a single king of expansion, but it is not true. E.g. Hadamard gradient has a different one.

File: pennylane/math/multi_dispatch.py [hunk 20]
```diff
@@ -898,6 +880,37 @@ def detach(tensor, like=None):
     return tensor
 
 
+def jax_argnums_to_tape_trainable(qnode, argnums, expand_fn, args, kwargs):
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/math/multi_dispatch.py:883]: I will update the docs thanks!

File: pennylane/transforms/classical_jacobian.py (hunk 18 above)
Comment (rmoyard) [reply pennylane/transforms/classical_jacobian.py:187]: It is a kwarg in the `classical_jacobian`, but yes the implementation was very tricky.

File: pennylane/math/multi_dispatch.py [hunk 21]
```diff
@@ -898,6 +880,37 @@ def detach(tensor, like=None):
...
+    args_jvp = [
+        jax.ad.JVPTracer(trace, arg, jax.numpy.zeros(1)) if i in argnums else arg
+        for i, arg in enumerate(args)
+    ]
+
+    qnode.construct(args_jvp, kwargs)
+    tape = qnode.qtape
+    tape = expand_fn(tape)
+    params = tape.get_parameters(trainable_only=False)
+    del trace
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/math/multi_dispatch.py:910]: Yes I think you are correct that it del when there are no references I think.

File: doc/development/deprecations.rst [hunk 22]
```diff
@@ -6,6 +6,12 @@ Deprecations
 Pending deprecations
 --------------------
 
+* The argument ``argnum`` for gradient transforms is replaced by ``argnums``.
```
Comment (rmoyard) [inline doc/development/deprecations.rst:9]:
```suggestion
* The argument ``argnum`` for gradient transforms using the Jax interface is replaced by ``argnums``.
```

**Review by timmysilv — COMMENTED** [review]
just some small comments left

File: pennylane/transforms/metric_tensor.py [hunk 23]
```diff
@@ -383,6 +383,8 @@ def qnode_execution_wrapper(self, qnode, targs, tkwargs):
     mt_fn = self.default_qnode_wrapper(qnode, targs, tkwargs)
 
     def _expand_fn(tape):
+        # Currently does not support it
+        tkwargs.pop("argnums", None)
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/transforms/metric_tensor.py:387]: https://github.com/PennyLaneAI/pennylane/pull/3847

File: tests/gradients/test_parameter_shift_hessian.py [hunk 24]
```diff
@@ -1328,7 +1328,7 @@ def circuit(x):
 
         expected = qml.jacobian(qml.jacobian(circuit))(x_np)
         circuit.interface = "jax"
-        hess = qml.gradients.param_shift_hessian(circuit)(x_jax)
+        hess = qml.gradients.param_shift_hessian(circuit, argnums=[0])(x_jax)
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply tests/gradients/test_parameter_shift_hessian.py:1331]: It is default yes :+1:

File: pennylane/gradients/gradient_transform.py [hunk 25]
```diff
@@ -373,6 +408,8 @@ def jacobian_wrapper(
                 jacs = tuple(
                     qml.math.tensordot(qjac, c, [[-1], [0]]) for c in cjac if c is not None
                 )
+                if len(jacs) == 1:
+                    return jacs[0]
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/gradients/gradient_transform.py:412]: Yes correct

File: pennylane/gradients/gradient_transform.py [hunk 26]
```diff
@@ -273,23 +273,44 @@ def __init__(
...
+
+            argnums_ = None
+            if interface == "jax" and not trainable_params:
+                if argnums is None:
+                    argnums_ = [0]
+
+                if argnums is not None:
+                    argnums_ = [argnums] if isinstance(argnums, int) else argnums
+
+                params = qml.math.jax_argnums_to_tape_trainable(
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/gradients/gradient_transform.py:307]: I think it is better to keep it separates, because it could be useful and what is happening is clear.

**Review by timmysilv — APPROVED** [review]
can please codecov/pylint in this PR or the other one, depending on when you want this to go to master. but looks great!!

File: pennylane/math/__init__.py [hunk 27]
```diff
@@ -61,6 +61,7 @@
     tensordot,
     unwrap,
     where,
+    jax_argnums_to_tape_trainable,
```
Comment (dwierichs) [inline pennylane/math/__init__.py:64]: I think thus far this tuple is sorted alphabetically, maybe good to keep it sorted :)

File: pennylane/math/multi_dispatch.py [hunk 28]
```diff
@@ -898,6 +880,37 @@ def detach(tensor, like=None):
...
+    Return:
+        list[float, jax.JVPTracer]: List of parameters where the trainable one are `JVPTracer`.
+    """
+    import jax
+
+    with jax.core.new_main(jax.ad.JVPTrace) as main:
+        trace = jax.ad.JVPTrace(main, 0)
+
+    args_jvp = [
+        jax.ad.JVPTracer(trace, arg, jax.numpy.zeros(1)) if i in argnums else arg
```
Comment (dwierichs) [inline pennylane/math/multi_dispatch.py:902]: > jax.numpy.zeros(1)

Does this assume scalar tape parameters?

File: tests/transforms/test_classical_jacobian.py [hunk 29]
```diff
@@ -544,3 +443,169 @@ def test_torch_not_trainable_only(diff_method, interface):
...
+            ]
+        ),
+        np.vstack([np.array([1, 0.4 * z[1], 0.0, 0.0])]),
+        tuple(
+            [
+                np.vstack([np.eye(len(x)), np.zeros((1, 3))]),
+                np.vstack([np.zeros((3, 4)), np.array([1, 0.4 * z[1], 0.0, 0.0])]),
+            ]
+        ),
+    ),
```
Comment (dwierichs) [inline tests/transforms/test_classical_jacobian.py:514]: Seems like a lot of line breaks have been made by black because of trailing commas, no? Many seem unnecessary.

**Review by dwierichs — COMMENTED** [review]
Some minor cosmetic suggestions, otherwise LGTM! :100: 
Those tests though :o 
I'm wondering whether we should not test all children classes of `gradient_transform` whenever we change some general behaviour of `gradient_transform`...

File: pennylane/math/multi_dispatch.py (hunk 28 above)
Comment (rmoyard) [reply pennylane/math/multi_dispatch.py:902]: Yes I have changed it in the next PR https://github.com/PennyLaneAI/pennylane/pull/3847

File: pennylane/math/__init__.py (hunk 27 above)
Comment (rmoyard) [reply pennylane/math/__init__.py:64]: Good catch!

File: tests/transforms/test_classical_jacobian.py (hunk 29 above)
Comment (timmysilv) [reply tests/transforms/test_classical_jacobian.py:514]: I thought black adds the commas _and_ breaks :p

Comment (rmoyard) [reply tests/transforms/test_classical_jacobian.py:514]: I think with that view it is easy to see that there are three elements in each tuple and identify them.

File: pennylane/gradients/gradient_transform.py [hunk 30]
```diff
@@ -303,7 +324,21 @@ def jacobian_wrapper(
...
+            # defined on the outer transform and therefore on the args.
+            if argnums is None and argnums_ is None and interface == "jax":
+                cjac = qml.transforms.classical_jacobian(
+                    qnode, argnum=qml.math.get_trainable_indices(args), expand_fn=self.expand_fn
+                )(*args, **kwargs)
+            else:
+                if interface == "jax":
+                    argnum_cjac = argnums if argnums is not None else [0]
+                else:
+                    argnum_cjac = None
```
Comment (timmysilv) [inline pennylane/gradients/gradient_transform.py:338]: doesn't this mean that `argnum_cjac` is always `None` for non-jax computations?

**Review by timmysilv — CHANGES_REQUESTED** [review]
I'd like to have [this comment from the metric_tensor PR](https://github.com/PennyLaneAI/pennylane/pull/3847#discussion_r1124814259) addressed before either PR is merged.

File: pennylane/gradients/gradient_transform.py (hunk 30 above)
Comment (rmoyard) [reply pennylane/gradients/gradient_transform.py:338]: Yes it is correct, it is the current way of working for other interfaces, you don't pass the argnum to cjac.

Comment (timmysilv) [reply pennylane/gradients/gradient_transform.py:338]: it kind of looks like we do [here](https://github.com/PennyLaneAI/pennylane/blob/b36054fbe7bf62f1186f14139d08c8d189a5f8cc/pennylane/gradients/gradient_transform.py#L286), no?

Comment (rmoyard) [reply pennylane/gradients/gradient_transform.py]: It is a recent change that I made for the Jax interface (as a first step to make it work), argnums is always None for other interfaces.

Comment (rmoyard) [reply pennylane/gradients/gradient_transform.py]: https://github.com/PennyLaneAI/pennylane/pull/3776/files#diff-7498bddbc39a804610097ec71b60020e226e1f244f3d4d124faa6be9cfdd1c91R285

Comment (rmoyard) [reply pennylane/gradients/gradient_transform.py]: Simplification done, thanks for pointing this out :+1:

**Review by timmysilv — APPROVED** [review]

File: pennylane/math/multi_dispatch.py (hunk 28 above)
Comment (dwierichs) [reply pennylane/math/multi_dispatch.py:902]: Great, thanks!

File: tests/transforms/test_classical_jacobian.py (hunk 29 above)
Comment (dwierichs) [reply tests/transforms/test_classical_jacobian.py:514]: Fair enough, don't have a strong opinion here. 

> I thought black adds the commas and breaks :p

Ah really? I thought it breaks if there is a comma and doesn't if not (if the line length is small enough)

Comment (dwierichs) [reply tests/transforms/test_classical_jacobian.py:514]: Yup, confirmed this on a tiny test file: 
```python
a = (1, 2, 3,)
```
is converted to 
```python
a = (
    1,
    2,
    3,
)
```
and 
```python
a = (1, 2, 3)
```
is left as is.
 @timmysilv
